src/expr/parser.rs

```diff
@@ -25,34 +26,85 @@ named!(i64_literal<i64>,
   )
 );
 
-named!(factor<i64>,
+named!(literal<Expr>,
+  map!(
+    isize_digit,
+    Expr::Literal
+  )
+);
+
+named!(factor<Expr>,
   alt!(
-    i64_literal
+    literal
   | parens
   )
 );
 
-// we define acc as mutable to update its value whenever a new term is found
-named!(term <i64>,
+named!(term <Expr>,
   chain!(
-    mut acc: factor  ~
+    mut acc: factor ~
              many0!(
                alt!(
-                 tap!(mul: preceded!(tag!("*"), factor) => acc = acc * mul) |
-                 tap!(div: preceded!(tag!("/"), factor) => acc = acc / div)
+                 map!(preceded!(tag!("*"), factor),
+                      |mul| { match acc {
+                          Expr::Literal(x) => acc =
+                              Expr::BinOp(Box::new(Expr::Literal(x)), BinOp::Times, Box::new(mul)),
+                          Expr::BinOp(ref mut left, ref mut op, ref mut right) => {
+                              let temp = Expr::Literal(0);
+                              let l = mem::replace(left, Box::new(temp));
+                              let r = mem::replace(right, Box::new(mul));
+                              let o = mem::replace(op, BinOp::Times);
+                              mem::replace(left, Box::new(Expr::BinOp(l, o, r)));
+                          }
+                      }}) |
```

Comment:
So this ends up working. You match on `acc`, and in the `Literal` case you bind the `i64` directly (which doesn't move `acc` because `i64: Copy`. In the `BinOp` case you bind all contents by mutable reference, you also avoid the move.

There is a more clean way though: you can make sure to refer to use `acc` only as a mutable referenceto begin with. This also avoids the move, and shortens your code a bunch. 

Check out the sample solution for an example of this.